roofit/roofitcore/src/TestStatistics/MinuitFcnGrad.cxx

```diff
@@ -0,0 +1,206 @@
+// Author: Patrick Bos, Netherlands eScience Center / NIKHEF 2021
+
+/*****************************************************************************
+ * RooFit
+ * Authors:                                                                  *
+ *   WV, Wouter Verkerke, UC Santa Barbara, verkerke@slac.stanford.edu       *
+ *   DK, David Kirkby,    UC Irvine,         dkirkby@uci.edu                 *
+ *                                                                           *
+ * Copyright (c) 2000-2021, Regents of the University of California          *
+ *                          and Stanford University. All rights reserved.    *
+ *                                                                           *
+ * Redistribution and use in source and binary forms,                        *
+ * with or without modification, are permitted according to the terms        *
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+
+#include "TestStatistics/MinuitFcnGrad.h"
+
+#include "RooMsgService.h"
+#include "RooAbsPdf.h"
+
+namespace RooFit {
+namespace TestStatistics {
+
+// IMultiGradFunction overrides necessary for Minuit: DoEval, Gradient
+// The likelihood and gradient wrappers do the actual calculations.
+
+double MinuitFcnGrad::DoEval(const double *x) const
+{
+   Bool_t parameters_changed = syncParameterValuesFromMinuitCalls(x, false);
+
+   // Calculate the function for these parameters
+//   RooAbsReal::setHideOffset(kFALSE);
+   likelihood->evaluate();
+   double fvalue = likelihood->getResult();
+   calculation_is_clean->likelihood = true;
+//   RooAbsReal::setHideOffset(kTRUE);
+
+   if (!parameters_changed) {
+      return fvalue;
+   }
+
+   if (!std::isfinite(fvalue) || RooAbsReal::numEvalErrors() > 0 || fvalue > 1e30) {
+
+      if (_printEvalErrors >= 0) {
+
+         if (_doEvalErrorWall) {
+            oocoutW(static_cast<RooAbsArg *>(nullptr), Eval)
+               << "RooGradMinimizerFcn: Minimized function has error status." << std::endl
+               << "Returning maximum FCN so far (" << _maxFCN
+               << ") to force MIGRAD to back out of this region. Error log follows" << std::endl;
+         } else {
+            oocoutW(static_cast<RooAbsArg *>(nullptr), Eval)
+               << "RooGradMinimizerFcn: Minimized function has error status but is ignored" << std::endl;
+         }
+
+         TIterator *iter = _floatParamList->createIterator();
```

Comment:
There shouldn't be `TIterator`s starting from ROOT 6.20. They are slower, less type safe, prone to memory leaks and harder to read. All these loops can usually be simplified using:
```diff
- TIterator *iter = _floatParamList->createIterator();
- RooRealVar *var;
- while ((var = (RooRealVar *)iter->Next())) {
-   ...
- }
- delete iter;
+ for (const auto rooAbsArg : _floatParamList) {
+   auto var = static_cast<const RooRealVar*>(rooAbsArg);
+  ...
+ }
```

Reply:
Thanks, done. Also for the three other remaining TIterators in TestStatistics, one in likelihood_builders.cxx, one in RooSubsidiaryL.cxx and one in ConstantTermsOptimizer.cxx (renamed from optimization.cxx as suggested).